core/src/main/scala/cats/Eval.scala

```diff
@@ -318,52 +318,61 @@ object Eval extends EvalInstances {
       }
   }
 
-  private def evaluate[A](e: Eval[A]): A = {
-    type L = Eval[Any]
-    type M = Memoize[Any]
-    type C = Any => Eval[Any]
+  /*
+   * This represents the stack of flatmap functions in a series
+   * of Eval operations
+   */
+  sealed abstract private class FnStack[-A, +B]
+  final private case class Ident[A, B](ev: A <:< B) extends FnStack[A, B]
+  final private case class Many[A, B, C](first: A => Eval[B], rest: FnStack[B, C]) extends FnStack[A, C]
 
-    def addToMemo(m: M): C = { (a: Any) =>
+  private def evaluate[A](e: Eval[A]): A = {
+    def addToMemo[A1](m: Memoize[A1]): A1 => Eval[A1] = { (a: A1) =>
       m.result = Some(a)
       Now(a)
     }
 
-    @tailrec def loop(curr: L, fs: List[C]): Any =
+    @tailrec def loop[A1](curr: Eval[A1], fs: FnStack[A1, A]): A =
       curr match {
-        case c: FlatMap[_] =>
+        case c: FlatMap[A1] =>
+          val nextFs: Many[c.Start, A1, A] = Many(c.run, fs)
           c.start() match {
-            case cc: FlatMap[_] =>
-              loop(cc.start().asInstanceOf[L], cc.run.asInstanceOf[C] :: c.run.asInstanceOf[C] :: fs)
+            case cc: FlatMap[c.Start] =>
+              loop(cc.start(), Many(cc.run, nextFs))
             case mm @ Memoize(eval) =>
               mm.result match {
                 case Some(a) =>
-                  loop(Now(a), c.run.asInstanceOf[C] :: fs)
+                  loop(nextFs.first(a), nextFs.rest)
                 case None =>
-                  loop(eval, addToMemo(mm.asInstanceOf[M]) :: c.run.asInstanceOf[C] :: fs)
+                  loop(eval, Many(addToMemo(mm), nextFs))
               }
             case xx =>
+              // note: xx could be a Defer so calling .value
+              // could blow the stack?
```

Comment:
this looks unsafe to me. I don't see why we don't need to check for `Defer` here. I think this could be the cause of #2587

If this is safe, I think we need a comment as to why `xx.value` is okay.

Reply:
@non do you know why we wouldn't do `loop(xx, Many(c.run, fs))` when we have a Defer?

Reply:
Okay, so `defer(e).flatMap(f)` is safe because `.flatMap` checks for `Defer` and uses the thunk as `start`, BUT `defer(defer(e)).flatMap(f)` isn't safe because flatMap only unwraps one layer of `Defer` and then we wind up calling `.value` on a defer which blows the stack.

I reproed this failure mode in a test.

Should I combine with this code for the fix or make a follow up PR?